cli/cmd/profile.go

```diff
@@ -0,0 +1,94 @@
+package cmd
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"text/template"
+
+	"github.com/linkerd/linkerd2/cli/profile"
+	"github.com/spf13/cobra"
+)
+
+type templateConfig struct {
+	ControlPlaneNamespace string
+	ServiceNamespace      string
+	ServiceName           string
+}
+
+func newCmdProfile() *cobra.Command {
+
+	var namespace string
+	var template bool
+
+	cmd := &cobra.Command{
+		Use:   "profile [flags] --template (SERVICE)",
+		Short: "Output template service profile config for Kubernetes",
+		Long: `Output template service profile config for Kubernetes.
+		
+		This outputs a service profile template for the given service.  Edit the
+		template and then apply it with kubectl to add a service profile to a
+		service.
+
+		Example:
+		linkerd profile --template web-svc.emojivoto > web-svc-profile.yaml
+		# (edit web-svc-profile.yaml manually)
+		kubectl apply -f web-svc-profile.yaml
+		`,
+		Args: cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			config, err := buildConfig(namespace, args[0])
+			if err != nil {
+				return err
+			}
+			return renderProfileTemplate(config, os.Stdout)
+		},
+	}
+
+	cmd.PersistentFlags().BoolVar(&template, "template", true, "Output a service profile template")
+	cmd.PersistentFlags().StringVarP(&namespace, "namespace", "n", namespace, "Namespace of the service")
+
+	return cmd
+}
+
+func buildConfig(namespace, service string) (templateConfig, error) {
+
+	serviceParts := strings.Split(service, ".")
+
+	serviceName := serviceParts[0]
+	serviceNamespace := "default"
+
+	if len(serviceParts) >= 2 && namespace != "" && serviceParts[1] != namespace {
+		return templateConfig{}, fmt.Errorf("service namespace cannot be specified by namespace flag (%s) and by service name (%s)", namespace, serviceParts[1])
```

Comment:
Ahh, regarding my comment about the example above -- my preference is to only support namespaces provided via the `--namespace` flag. That will be more consistent with how we treat stat and tap targets. I'd just simplify this to:

```diff
diff --git a/cli/cmd/profile.go b/cli/cmd/profile.go
index 93244a30..ca2cad37 100644
--- a/cli/cmd/profile.go
+++ b/cli/cmd/profile.go
@@ -2,10 +2,8 @@ package cmd
 
 import (
 	"bytes"
-	"fmt"
 	"io"
 	"os"
-	"strings"
 	"text/template"
 
 	"github.com/linkerd/linkerd2/cli/profile"
@@ -39,11 +37,7 @@ func newCmdProfile() *cobra.Command {
 		`,
 		Args: cobra.ExactArgs(1),
 		RunE: func(cmd *cobra.Command, args []string) error {
-			config, err := buildConfig(namespace, args[0])
-			if err != nil {
-				return err
-			}
-			return renderProfileTemplate(config, os.Stdout)
+			return renderProfileTemplate(buildConfig(namespace, args[0]), os.Stdout)
 		},
 	}
 
@@ -53,32 +47,15 @@ func newCmdProfile() *cobra.Command {
 	return cmd
 }
 
-func buildConfig(namespace, service string) (templateConfig, error) {
-
-	serviceParts := strings.Split(service, ".")
-
-	serviceName := serviceParts[0]
-	serviceNamespace := "default"
-
-	if len(serviceParts) >= 2 && namespace != "" && serviceParts[1] != namespace {
-		return templateConfig{}, fmt.Errorf("service namespace cannot be specified by namespace flag (%s) and by service name (%s)", namespace, serviceParts[1])
-	}
-
-	if len(serviceParts) >= 2 {
-		serviceNamespace = serviceParts[1]
-	}
-
-	if namespace != "" {
-		serviceNamespace = namespace
+func buildConfig(namespace, service string) *templateConfig {
+	return &templateConfig{
+		ControlPlaneNamespace: controlPlaneNamespace,
+		ServiceNamespace:      namespace,
+		ServiceName:           service,
 	}
-	return templateConfig{
-		controlPlaneNamespace,
-		serviceNamespace,
-		serviceName,
-	}, nil
 }
 
-func renderProfileTemplate(config templateConfig, w io.Writer) error {
+func renderProfileTemplate(config *templateConfig, w io.Writer) error {
 	template, err := template.New("profile").Parse(profile.Template)
 	if err != nil {
 		return err
```

(note that the diff also modifies this method to a return a pointer to the template struct, rather than the struct itself)

Reply:
just for my education, why is returning a pointer to the struct preferred over returning the struct directly?